synapse/storage/data_stores/main/events_worker.py

```diff
@@ -718,6 +718,9 @@ def _maybe_redact_event_row(self, original_ev, redactions, event_map):
             redacted_event = prune_event(original_ev)
             redacted_event.unsigned["redacted_by"] = redaction_id
 
+            # Remove this event from future search results
+            yield self.remove_event_from_event_search(original_ev.event_id)
```

Comment:
this will be called every time the redacted event is pulled out of the database. That feels like the wrong time to be updating event_search. In particular, this code is run on workers, which don't have a SearchStore (and shouldn't be writing to event_search, really).

I feel like we either need to update the SearchStore when we receive the redaction.

Reply:
... or check for redacted events when we process the results of a search.

Reply:
I first attempted to do the latter, and while we do when pulling the list of events, we can't easily do so when calculating the total count of events (because we only pull out a page of events at a time, and that's when we check if they're redacted events or not.

So doing it when we actually redact an event sounds good.

Reply:
Well, it's tough to actually do this as it is possible to get redactions of an event before the event itself actually comes in, so instead we'll want to do the check when a search occurs.

I tried this originally, but there was a problem with being able to get an accurate count of events as the counting part just uses the `event_search` table, which doesn't have any info about whether an event is redacted in it.

Instead, I'm going to try and come up with some SQL that'll let me query an accurate count.

Reply:
As [per #synapse-dev](https://matrix.to/#/!XaqDhxuTIlvldquJaV:matrix.org/$e-M7bzr_SFwAa_u2BaNd53Ffa3sVXI9ATWUWsArxd8w?via=matrix.org&via=amorgan.xyz&via=matrix.f5.htw-berlin.de), we'll just leave the inaccurate count for now and focus on the search results.